airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/format/containers.py

```diff
@@ -0,0 +1,125 @@
+
+    # Add any environment variables
+    for key, value in env_vars.items():
+        container = container.with_env_variable(key, value)
+
+    # Install any dependencies of the formatter
+    if install_commands:
+        container = container.with_exec(sh_dash_c(install_commands), skip_entrypoint=True)
+
+    # Mount the relevant parts of the repository: the code to format and the formatting config
+    # Exclude the default ignore list to keep things as small as possible
+    container = container.with_mounted_directory(
+        "/src",
+        dagger_client.host().directory(
+            ".",
+            include=include,
+            exclude=DEFAULT_FORMAT_IGNORE_LIST,
+        ),
+    )
+
+    # Set the working directory to the code to format
+    container = container.with_workdir("/src")
+
+    return container
+
+
+def format_java_container(ctx: ClickPipelineContext) -> dagger.Container:
+    """Format java, groovy, and sql code via spotless."""
+    return build_container(
+        ctx,
+        base_image=AMAZONCORRETTO_IMAGE,
+        include=[
+            "**/*.java",
+            "**/*.sql",
+            "**/*.gradle",
+            "gradlew",
+            "gradlew.bat",
+            "gradle",
+            "**/deps.toml",
+            "**/gradle.properties",
+            "**/version.properties",
+            "tools/gradle/codestyle/java-google-style.xml",
+            "tools/gradle/codestyle/sql-dbeaver.properties",
+        ],
+        install_commands=[
+            "yum update -y",
+            "yum install -y findutils",  # gradle requires xargs, which is shipped in findutils.
+            "yum clean all",
+        ],
+    )
+
+
+def format_js_container(ctx: ClickPipelineContext) -> dagger.Container:
+    """Format yaml and json code via prettier."""
+    return build_container(
+        ctx,
+        base_image=NODE_IMAGE,
+        include=["**/*.yaml", "**/*.yml", "**.*/json", "package.json", "package-lock.json"],
+        install_commands=["npm install -g npm@10.1.0", "npm install -g prettier@2.8.1"],
+    )
+
+
+def format_license_container(ctx: ClickPipelineContext, license_file: str) -> dagger.Container:
+    return build_container(
+        ctx,
+        base_image=GO_IMAGE,
+        include=["**/*.java", "**/*.py", license_file],
+        install_commands=["go get -u github.com/google/addlicense"],
+    )
+
+
+def format_python_container(ctx: ClickPipelineContext) -> dagger.Container:
+    """Format python code via black and isort."""
+
+    # Here's approximately what it would look like if we built it the other way. Doesn't
+    # quite work. Not sure if its putting more work into.
+
+    # base_container = with_python_base(ctx._click_context, python_version="3.10.13")
+    # container = with_installed_pipx_package(ctx._click_context, base_container, "poetry")
+    # container = mount_repo_for_formatting(container, include=["**/*.py", "pyproject.toml", "poetry.lock"])
+    # return container
```

Comment:
this was a thought experiment re: building the containers not in a shared way, but in the "building block" way we do it elsewhere in the package

Reply:
This isn't a bad pattern! Only the topmost layers are worth sharing anyway: the ones with the base images followed by apt-get whatever and pip install whatever.